apps/wing/src/commands/test.ts

```diff
@@ -91,12 +119,155 @@ async function testOne(entrypoint: string) {
     }
   }
 
+  return hasFailures;
+};
+
+async function testSimulator(synthDir: string) {
+  const s = new sdk.testing.Simulator({ simfile: synthDir });
+  await s.start();
+
+  const testRunner = s.getResource("root/cloud.TestRunner") as ITestRunnerClient;
+  const tests = await testRunner.listTests();
+  const filteredTests = pickOneTestPerEnvironment(tests);
+  const results = new Array<sdk.cloud.TestResult>();
+
+  // TODO: run these tests in parallel
+  for (const path of filteredTests) {
+    results.push(await testRunner.runTest(path));
+  }
+
+  await s.stop();
+
+  const hasFailures = printTestReport(synthDir, results);
+  if (hasFailures) {
+    process.exit(1);
+  }
+}
+
+async function testTfAws(synthDir: string): Promise<sdk.cloud.TestResult[]> {
+  if (!isTerraformInstalled(synthDir)) {
+    throw new Error(
+      "Terraform is not installed. Please install Terraform to run tests in the cloud."
+    );
+  }
+
+  await withSpinner("terraform init", () => terraformInit(synthDir));
+
+  await checkTerraformStateIsEmpty(synthDir);
+
+  await withSpinner("terraform apply", () => terraformApply(synthDir));
+
+  const [testRunner, tests] = await withSpinner("Setting up test runner...", async () => {
+    const testArns = await terraformOutput(synthDir, "WING_TEST_RUNNER_FUNCTION_ARNS");
+    const testRunner = new TfawsTestRunnerClient(testArns);
+
+    const tests = await testRunner.listTests();
+    return [testRunner, pickOneTestPerEnvironment(tests)];
+  });
+
+  const results = await withSpinner("Running tests...", async () => {
+    const results = new Array<sdk.cloud.TestResult>();
+    for (const path of tests) {
+      results.push(await testRunner.runTest(path));
+    }
+    return results;
+  });
+
+  const hasFailures = printTestReport(synthDir, results);
+
   if (hasFailures) {
```

Comment:
> How do they cleanup the resources in the interim? Is the experience that we have them run `terraform destroy` in different directories? (until we add the `--destroy` option)

Yeah, that would the experience at the moment. cd into `target/test/<your-app-path>` and then `terraform destroy`.

> Since the default way to manage TF state is local state files, what if the dev runs into a scenario where their local state file gets blown to bits or something. Does this just orphan resources in their account that they need to cleanup?

If you deleted your `target/` directory or otherwise tampered with the contents, then I think resources would be orphaned, yes. (I think losing state files is a limitation/risk inherent to terraform if I understand correctly). But as long as you're just using Wing CLI commands or the Wing Console that shouldn't happen. I think it'll probably be possible to use have Wing use remote terraform state via a Wing plugin.

> I wonder if we should by default always cleanup and make an optional flag --retain-on-fail or something so that test resources are not accidentally left over in an account, to avoid anyone getting a suprise AWS bill. I think my mental model of "test" resources are that they should be always be ephemeral unless someone presses the "I know what Im doing button"

I'm empathetic to the idea of cleaning up by default because of the high costs of resources in the cloud, but I guess I also just worry about what the developer experience would be like if a test failed and it automatically started cleaning it up before it even asked you (because that means you might have to way for an entire deployment cycle in order to reproduce the situation). Maybe we could have it so as long as you're running `wing test` interactively, we can prompt you whether you want to retain or not, and if it's not interactive the CLI will clean up by default (or retain resources if `--retain-on-fail` was passed)?

Reply:
Yea I mean I think this maybe a philosophical choice. "To be cheap or to be convenient" 😂 I totally see the reasoning for both arguments. I know a lot of people get frustrated with cloud formation deployments since the default behavior is "If it failed roll it back" but personally I prefer this. 

When it comes to a developer experience, I feel like the time cost of having to re-run with a `--retain-on-failure` is preferable to "oops I left 8 copies of the app running last month" kinda thing.

Reply:
I've switched it clean up by default, resolving for now (let's be sure to dogfood this to make sure this default feels right)